src/com/linkedin/parseq/FusionTask.java

```diff
@@ -22,33 +22,61 @@
 
 
 /**
+ * FusionTask implements optimization in case when multiple synchronous
+ * transformations are applied in sequence. Instead of rescheduling every transformation
+ * as a separate task it is faster to apply transformations in a chain.
  *
- * @author jodzga
+ * This is internal class and it should not be extended or used outside ParSeq.
+ * It may change or be removed at any time in a backwards incompatible way.
  *
- * @param <S>
- * @param <T>
+ * @author Jaroslaw Odzga (jodzga@linkedin.com)
+ *
+ * @param <S> type of source
+ * @param <T> type of target
  */
-public class FusionTask<S, T> extends BaseTask<T> {
+class FusionTask<S, T> extends BaseTask<T> {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(FusionTask.class);
 
   private static final Continuations CONTINUATIONS = new Continuations();
 
+  /* Encapsulates transformation from source to target, includes all predecessors' transformations, can't be null */
   private final Consumer3<FusionTraceContext, Promise<S>, Settable<T>> _propagator;
-  private final Task<S> _task;
+
+  /* Asynchronous task that will complete source promise, can be null in which case propagator must not depend on source */
+  private final Task<S> _asyncTask;
+
+  /* Source that will be used by propagator, can be null in which case propagator must not depend on it */
   private final Promise<S> _source;
 
+  /* Trace builder for the predecessor, used for building trace relationships, can be null */
   private final ShallowTraceBuilder _predecessor;
 
   private FusionTask(final String desc, final Task<S> task, final Promise<S> source,
-      final Consumer3<FusionTraceContext, Promise<S>, Settable<T>> propagator, final boolean systemHidden,
-      final ShallowTraceBuilder predecessor) {
+      final PromisePropagator<S, T> propagator) {
     super(desc);
-    _shallowTraceBuilder.setSystemHidden(systemHidden);
-    _propagator = completing(propagator);
-    _task = task;
+    _propagator = completing(adaptToAcceptTraceContext(propagator));
+    _asyncTask = task;
     _source = source;
-    _predecessor = predecessor;
+    _predecessor = null;
+  }
+
+  private <R> FusionTask(final String desc, final FusionTask<S, R> predecessor,
+      final PromisePropagator<R, T> propagator) {
+    super(desc);
+    _asyncTask = predecessor._asyncTask;
+    _source = predecessor._source;
+    _predecessor = predecessor.getShallowTraceBuilder();
```

Comment:
maybe rename _predecessor to _predecessorShallowTraceBuilder... it's kind of long though